fix(orderbook): validate roots before commit

shamardy opened the pull request:
### Background
- After the P2P upgrade in #1878 (“broadcast one topic at a time”), keep‑alive messages no longer implied deletion when a pair’s root was absent. Combined with a single pubkey‑wide timeout, a seed that missed a cancel could keep an order until the maker went offline.
- After #2580, older nodes still didn't clear local pairs on null roots, leading to bad orderbooks and repeated warnings like “Couldn’t find an order … it will be synced upon pubkey keep alive,” and errors such as `InvalidStateRoot([0,…])` (see #2594).
- This PR finalizes the receiver‑side handling for the #2580 change by:
  - Validating per‑pair diffs strictly against maker‑advertised roots and clearing pairs on mismatch.
  - Tracking per‑pair liveness locally (`pair_last_seen_local`) and pruning stale pairs, removing a pubkey only when all its pairs are stale.

### What this PR changes
- Commit-after-validate (per pair)
  - On `PubkeyKeepAlive`, compute expected roots per pair from the maker.
  - Request diffs/full tries only from the keep‑alive’s origin peer (`propagated_from`).
  - Ignore unsolicited pairs in the sync response.
  - Apply the diff/full trie and recompute the root per pair; commit only if it exactly matches the expected root.
  - On mismatch, reject the data, clear that pair locally, and do not propagate the message.
  - The same “do not propagate” rule applies to stale/replayed keep‑alives handled as `StaleKeepAlive`.
   - Propagation: forward the keep‑alive only if all requested pairs were validated and applied successfully; if any pair is stale, mismatched, or remains unresolved, the message is not propagated.
- Unresolved pairs
  - If any requested pairs remain unresolved after the origin‑peer sync, return a `SyncFailure` (treated as a warning) and do not propagate the message.
- Stale/replay protection (per pair)
  - Introduces a per‑pair monotonic maker timestamp gate (`latest_root_timestamp_by_pair`) and a local per‑pair last‑seen clock (`pair_last_seen_local`).
  - Rejects stale keep‑alives with `OrderbookP2PHandlerError::StaleKeepAlive`.
  - Replay guard: we intentionally retain `latest_root_timestamp_by_pair` entries even after a pair is pruned, to block stale replays of old roots; these entries are dropped only when the entire pubkey state is removed (which is not ideal still).
- Receiver‑side handling of empty roots
  - If a keep‑alive carries a zero or hashed‑null root for a pair, clear local state for that pair and:
    - Store the null root in `trie_roots`,
    - Update the per‑pair maker timestamp (`latest_root_timestamp_by_pair`),
    - Remove the local last‑seen (`pair_last_seen_local`) so the pair becomes eligible for GC if it stays inactive.
  - No sync is attempted for that pair.
- Liveness and GC
  - Track liveness per pair and prune only stale pairs; remove the pubkey when all its pairs are stale.
  - GC now uses the constant `MAKER_ORDER_TIMEOUT`; the loop no longer reads an override from config.
  - `GetOrderbookPubkeyItem.last_keep_alive` reflects the max local last‑seen across the pubkey’s pairs.
  - When the orderbook is filled from relays (`GetOrderbook` response), we now set `pair_last_seen_local` for each imported (pubkey, pair), so `last_keep_alive` and GC behave correctly for imported state.
- Errors and logging
  - Adds `OrderbookP2PHandlerError::StaleKeepAlive` and `::SyncFailure` and treats `SyncFailure` as a warning to accommodate outdated peers.
- Code cleanup:
   - Removed a dead condition in keep‑alive broadcasting where `if (root == H64::default()) && (root == hashed_null_node::<Layout>())` can never be true (no functional change)
      - Maybe we should not broadcast such roots and let them timeout on receiver side instead to reduce network load, will think of that.

### Why this stops “ghost orders”
- We accept per‑pair data only if it reproduces the maker’s expected root.
- We time out inactive pairs locally (per pair) using the node’s clock.

### Out of scope (future work)
- Peer scoring/quarantine for consistently failing relays.

### TODO
- [ ] Maybe we should not broadcast null roots and let them timeout on receiver side instead to reduce network load.
- [ ] Refactor keepalive logic / code to a different file to start reducing the code in `lp_ordermatch.rs`. This will be done after reviews and approvals.

addresses #2594

shamardy committed:
ordermatch: commit-after-validate, sequential per-peer sync, synced flag/logs
- On non-null roots, sync only if the applied diff/full trie reproduces the maker-advertised expected root; otherwise revert the pair and continue to the next peer.
- Walk peers sequentially (origin first); mark pubkeys unsynced on unresolved pairs and log consulted peers.

shamardy committed:
remove unreachable null-root guard
- `H64::default()`` is `[0;8]`` while `hashed_null_node::<Layout>()` is `Blake2b-8(0x00) ≠ 0`, so the && condition could never be true.

shamardy committed:
enforce global maker timeout for remote pubkeys
- Remove per-node `maker_order_timeout` override; use `MAKER_ORDER_TIMEOUT` so remote order GC is consistent across the network

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2595,6 +2745,7 @@ struct OrderbookPubkeyState {
     orders_uuids: HashSet<(Uuid, AlbOrderedOrderbookPair)>,
     /// The map storing alphabetically ordered pair with trie root hash of orders owned by pubkey.
     trie_roots: HashMap<AlbOrderedOrderbookPair, H64>,
+    is_synced: bool,
```
let's add a doc comment here

shamardy replied:
Not relevant anymore

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2598,10 +2598,6 @@ fn broadcast_keep_alive_for_pub(ctx: &MmArc, pubkey: &str, orderbook: &Orderbook
     };
 
     for (alb_pair, root) in state.trie_roots.iter() {
-        if *root == H64::default() && *root == hashed_null_node::<Layout>() {
-            continue;
-        }
```
i think it was meant to be or-ed like in: https://github.com/KomodoPlatform/komodo-defi-framework/blob/b2217f67cdf4c9b9cb37e16a491af4d02c55e5ca/mm2src/mm2_main/src/lp_ordermatch.rs#L3111-L3119

shamardy replied:
Yep, but I decided to broadcast null roots for now, will reassess this though after I decided to timeout maker orders per pair.

shamardy replied:
Added as a Todo here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605#issue-3339433732, will resolve this comment now

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -310,50 +312,202 @@ fn process_trie_delta(
+    };
+    let mut remaining_pairs: HashSet<AlbOrderedOrderbookPair> = expected_roots_by_pair.keys().cloned().collect();
+
+    // Build the peer sequence once: origin first, then current mesh (excluding origin).
```
let's remove `once`. it's confusing.

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -310,50 +312,202 @@ fn process_trie_delta(
+    let peers = {
+        let p2p_cmd_tx = P2PContext::fetch_from_mm_arc(&ctx).cmd_tx.lock().clone();
+        let mut mesh = get_relay_mesh(p2p_cmd_tx).await;
+        mesh.retain(|p| p != &propagated_from);
```
i think we better use something like swap_remove (if we don't care about the order) to avoid elements being moved inside the vector.

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -310,50 +312,202 @@ fn process_trie_delta(
+        let p2p_cmd_tx = P2PContext::fetch_from_mm_arc(&ctx).cmd_tx.lock().clone();
+        let mut mesh = get_relay_mesh(p2p_cmd_tx).await;
+        mesh.retain(|p| p != &propagated_from);
+        std::iter::once(propagated_from.clone())
```
what do we mean by `origin`, is the the message original sender or the peer that forwarded it to us?
isn't `propagated_from` is the peer that forward the message to us?

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -310,50 +312,202 @@ fn process_trie_delta(
+            &expected_roots_by_pair,
+            &mut remaining_pairs,
+        )
+        .await?;
```
what about ignoring the error? i guess we wanna perform these requests in a best-effort basis. if we fail to get a response (idk how fail cases happen since `PeerDecodedResponse` is not really well-documented :/), we can try the other peers we have in the list.

shamardy replied:
Will not ignore this error in the next commit

shamardy replied:
Nevermind, will ignore it as if any error is returned from `request_and_apply_pubkey_state_sync_from_peer` `remaining_pairs` will not be empty and it will be a `SyncFailure`

mariocynicys reviewed (commented):
first iteration. a little messy in my head. still attaching strings.

1- could we add more doc comments for new functions.

2- so the previous code used to request the updated orderbook for maker x from the propagator. now we ask the propagator and fall back to the entire mesh if the propagator sends invalid data.
the question is: why would the propagator ever send invalid or not-up-to-date data?!
since they sent/forwarded the keepalive message, doesn't this mean they processed it themselves and they have what we believe is the correct state that we call here `expected_roots_by_pair`?

i.e. `expected_roots_by_pair` is directly what the propagator has as their view of the roots for this maker. why could they send invalid data when we ask about the diff we are missing to reach this expected view/roots (that they themselves sent us).

shamardy committed:
wip

shamardy committed:
Merge remote-tracking branch 'origin/dev' into hotfix/sync-orderbook-fallback

dimxy commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -3714,7 +3874,6 @@ pub async fn lp_ordermatch_loop(ctx: MmArc) {
         .expect("CryptoCtx not available")
         .mm2_internal_pubkey_hex();
 
-    let maker_order_timeout = ctx.conf["maker_order_timeout"].as_u64().unwrap_or(MAKER_ORDER_TIMEOUT);
```
Why is `maker_order_timeout` not used now, to avoid short-lived orders?
I can see this param is used in tests, maybe could be enabled with "for-tests" feature?

shamardy replied:
Thanks for finding this, there is a test failing due to this removal. Will feature gate this to one of the test features / cfg for sure.

shamardy replied:
Decided to extend the wait in the failing test to 16 seconds instead here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605/commits/0f47c680c1e3432dbe3a20003c160def8d179427 since the timeout in tests is only 15 seconds.
https://github.com/KomodoPlatform/komodo-defi-framework/blob/e8372dc76d50110d1a90107b83618e6965f1ba21/mm2src/mm2_main/src/lp_ordermatch.rs#L139-L140
https://github.com/KomodoPlatform/komodo-defi-framework/blob/e8372dc76d50110d1a90107b83618e6965f1ba21/mm2src/mm2_main/src/lp_ordermatch.rs#L142

shamardy replied:
Test still failing. It might be a different problem other than timeout config in tests, will investigate it now.

shamardy replied:
Ok, `mm2_tests_main` doesn't obey `#[cfg(test)]` while `for-tests` feature requires changes in the CI when running `mm2_tests_main`, since `MIN_ORDER_KEEP_ALIVE_INTERVAL` is in the same crate as `mm2_tests_main`.

shamardy replied:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605/commits/84468c215b7703f46982773d0178ba7de512bbc4 should allow `for-tests` to be used inside `mm2_main` crate while wiring it in the right way. Any integration / docker test now requires `for-tests` / `run-docker-tests` feature to be passed, if not passed a compilation error will show. For docker tests ,`run-docker-tests` includes `for-tests` as a dependent feature.

I also added an explicit `[[test]]` targets which allowed running any test from the IDE using the “Run” (play) button to work on `mm2_tests_main` or `docker_tests_main`.

The failing `test_order_should_not_be_displayed_when_node_is_down` should now work by using the test value of `MIN_ORDER_KEEP_ALIVE_INTERVAL` as `for-tests` feature was added to it.

shamardy committed:
don't propogate stale PubkeyKeepAlives, log-and-continue on sync errors instead of aborting

dimxy commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2922,19 +3091,46 @@ impl Orderbook {
         i_am_relay: bool,
-    ) -> Option<OrdermatchRequest> {
+        propagated_from: &str,
+    ) -> Result<Option<OrdermatchRequest>, MmError<OrderbookP2PHandlerError>> {
```
Now the behaviour of the process_keep_alive fn seems to me not quite consistent:
In this fn we create OrdermatchRequest but on return from it we decompose this request to extract the trie_roots (and later recreate the request again). Would it not be better to just return the roots?

I think it's good to add a doc comment to the process_keep_alive fn.

shamardy replied:
> Would it not be better to just return the roots?

I guess so. Will fix it.

shamardy replied:
Done here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605/commits/89edefd99615c603fecd398a119178d92954e9f6

Please note that `process_orders_keep_alive` was broken down and `remaining_pairs` set was added to try to sync from another peer if validation failed for the `propagated_from` peer. I have rolled back this behaviour later due to problems that showed while @cipig tested, so I might even roll back parts of this refactor to make the changes more easy to review if it makes sense.

shamardy committed:
ordermatch: per-pair liveness, origin-only sync, pair-level GC

Track liveness per trading pair using maker‑published latest_root_timestamp_by_pair and local pair_last_seen_local. Keep‑alive processing now requests trie diffs only from the peer that propagated the message (origin‑only sync); if any pairs remain unresolved after syncing, the keep‑alive is not forwarded and a SyncFailure is logged (warn). Apply garbage collection at the pair level based on pair_last_seen_local, and remove the pubkey once all of its pairs are stale. Remove the legacy global last_keep_alive/latest_maker_timestamp fields and the is_synced flag.

shamardy committed:
Merge remote-tracking branch 'origin/dev' into hotfix/sync-orderbook-fallback

shamardy commented:
I guess this can be considered ready for review even though I still plan to optimize some things and improve the logs a bit.

onur-ozkan commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -176,11 +176,31 @@ pub enum OrderbookP2PHandlerError {
-        matches!(self, OrderbookP2PHandlerError::OrderNotFound(_))
+        matches!(
+            self,
+            OrderbookP2PHandlerError::OrderNotFound(_) | OrderbookP2PHandlerError::SyncFailure { .. } // treat SyncFailure as a warning for now due to outdated nodes
```
```suggestion
            // treat SyncFailure as a warning for now due to outdated nodes
            OrderbookP2PHandlerError::OrderNotFound(_) | OrderbookP2PHandlerError::SyncFailure { .. }
```

onur-ozkan commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -310,50 +330,182 @@ fn process_trie_delta(
+        if let Some(root) = expected_pair_roots.get(pair) {
+            trie_roots.insert(pair.clone(), *root);
+        }
+    }
```
```suggestion
    let trie_roots = pending_pairs
        .iter()
        .filter_map(|p| expected_pair_roots.get(p).map(|&root| (p.clone(), root)))
        .collect();
```

onur-ozkan commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2917,25 +3077,48 @@ impl Orderbook {
+    // For keep-alive, enforce a minimum interval per pubkey (e.g., >= 20–30s) or X messages/minute.
+    // When throttled, early-return Ok(None) to avoid sync and propagation.
+    // Prefer implementing this in a shared limiter (e.g., mm2_p2p or a small service in lp_network)
+    // and call it here before any state mutation.
```
I guess centralized and shared doesn't mean a http service that is deployed in a centralized server, or is it?

Because individual rate-limiting on seed-nodes should do the work.

shamardy replied:
> I guess centralized and shared doesn't mean a http service that is deployed in a centralized server, or is it?

Of course not, by “centralized” I meant a single, shared in-process limiter (module/utility), not any external service. I’ll reword the comment to make that clear.

shamardy replied:
Hope it's clearer here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605/commits/b009b53d6e8b53547f426b28f8ee52e8dd131c25

onur-ozkan commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -3739,26 +3931,78 @@ pub async fn lp_ordermatch_loop(ctx: MmArc) {
-            let mut uuids_to_remove = vec![];
-            let mut pubkeys_to_remove = vec![];
+
+            let deadline = now_sec().saturating_sub(MAKER_ORDER_TIMEOUT);
+
+            let mut pairs_to_prune = Vec::new();
+            let mut pubkeys_to_remove = Vec::new();
```
I am not entirely sure if it makes sense but I got an idea about doing a pre-check, something like `if deadline is hit even by the most recent order {}` and prune them all in a faster way without doing all this conditional work below.

shamardy replied:
Isn't the below equivalent to `if deadline is hit even by the most recent order {}`
https://github.com/KomodoPlatform/komodo-defi-framework/blob/b009b53d6e8b53547f426b28f8ee52e8dd131c25/mm2src/mm2_main/src/lp_ordermatch.rs#L3960-L3968

onur-ozkan replied:
It is, yes!

onur-ozkan commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -3739,26 +3931,78 @@ pub async fn lp_ordermatch_loop(ctx: MmArc) {
+                // - Cleanup: pre-scan without creating state; centralize timestamp checks/logging; never sync
+                //   origin for messages rejected by guards; add unit tests.
                 orderbook.pubkeys_state.remove(&pubkey);
             }
```
We are doubling the iteration here :)

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605/commits/ef4197c7876a08449a990ceaa0ee20423264e220

onur-ozkan reviewed (commented):
This is a rough review from my side. I will need to dig into the details later as I don't know the full context yet.

shamardy commented:
>  I will need to dig into the details later as I don't know the full context yet.

@onur-ozkan the description wasn't updated after the latest changes, done now. This should give you more context.

shamardy committed:
extend `test_order_should_not_be_displayed_when_node_is_down` wait to 16s.

shamardy committed:
feature‑gate integration tests in the right way
- for‑tests are now wired through the imported crates and  run‑docker‑tests includes it
- explicit [[test]] targets are declared with the required features
- order keep‑alive interval uses for-tests feature
- CI is updated to pass the for-test feature for integration tests

shamardy commented:
Please note that this commit https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605/commits/84468c215b7703f46982773d0178ba7de512bbc4 should be reviewed independent from all others ref. https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605#discussion_r2317512631

shamardy committed:
try to fix flaky `set_price_with_cancel_previous_should_broadcast_cancelled_message`
- It worked locally but failed in CI
- Now it uses `wait_for_log` instead of just sleeping

shamardy committed:
review fix: return per‑pair trie_roots from process_keep_alive and document behavior

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -808,8 +945,10 @@ fn process_get_orderbook_request(ctx: MmArc, base: String, rel: String) -> Resul
                 pubkey
             ))?;
 
+            let pubkey_last_seen = pubkey_state.pair_last_seen_local.values().copied().max().unwrap_or(0);
```
let's do `values().max().copied()` instead of `values().copied().max()` to not copy the whole iterator.

mariocynicys commented on mm2src/mm2_main/tests/feature_gate_for_tests.rs:
```diff
@@ -0,0 +1,8 @@
+#![cfg(not(target_arch = "wasm32"))]
+
+#[cfg(not(any(feature = "for-tests", feature = "run-docker-tests")))]
+compile_error!(concat!(
+    "Integration tests for `",
+    env!("CARGO_PKG_NAME"),
+    "` require either the `for-tests` or the `run-docker-tests` feature."
+));
```
do we need this now that we already have `required-features` in `Cargo.toml`?

shamardy replied:
This is needed when running individual tests

mariocynicys replied:
you mean individual tests like:  `-- test_sign_raw_transaction_rick` selection?
it still fails just fine locally :/
<img width="2576" height="524" alt="image" src="https://github.com/user-attachments/assets/b8070c09-c131-46d9-a05e-f8ae4cd3dbb3" />

shamardy replied:
yep, this compiler error is supposed to happen when not using the feature. Without it, the test will never run.

mariocynicys replied:
yup, that's relevant when running `cargo test` without test module specification (i.e. without `--test mm2_tests_main`)

mariocynicys commented on mm2src/mm2_main/tests/mm2_tests/mm2_tests_inner.rs:
```diff
@@ -2204,9 +2202,13 @@ fn set_price_with_cancel_previous_should_broadcast_cancelled_message() {
-    log!("Waiting ({} seconds) for Bob to broadcast messages…", pause);
-    thread::sleep(Duration::from_secs(pause));
+    block_on(mm_bob.wait_for_log(10., |log| log.contains("maker_order_cancelled_p2p_notify called")))
+        .expect("Cancel broadcast not seen in Bob’s logs within the expected time");
```
nit:
in `not seen in Bob’s logs`, in `Bob’s`, let's use ascii `'`

onur-ozkan reviewed (dismissed):
> > I will need to dig into the details later as I don't know the full context yet.
> 
> @onur-ozkan the description wasn't updated after the latest changes, done now. This should give you more context.

Thanks!

LGTM other than my previous review.

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -310,50 +330,164 @@ fn process_trie_delta(
+) {
+    let mut orderbook = orderbook_mutex.lock();
+    for (pair, diff) in response.pair_orders_diff {
+        // Ignore unsolicited pairs we didn't request to prevent state poisoning.
+        if !pending_pairs.contains(&pair) {
+            continue;
+        }
```
is that relevant anymore given that we only ask the source/origin peer (maker) for their orders?
they wouldn't poison our state with outdated orders.

shamardy replied:
We ask the seednode that propagated the keep alive message not the maker.

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -310,50 +330,164 @@ fn process_trie_delta(
+                    .insert(pair.clone(), keep_alive_timestamp);
+                state.pair_last_seen_local.insert(pair.clone(), now_sec());
+            } else {
+                warn!(
+                    "Sync validation failed for pubkey {} pair {} from {}: expected {:?}, got {:?}. Reverting pair.",
+                    from_pubkey, pair, peer, expected, new_root
+                );
+                remove_pubkey_pair_orders(&mut orderbook, from_pubkey, &pair);
+            }
```
the thing is, between a keepalive and the response from the source, some orders might have been added or removed that will change the `trie_root`, and we will purge the pair all together here.
i know we will re-construct it shortly, but since we now depend only on the origin (order publisher) as our source of truth, then we shouldn't be afraid of him lying to us (he wouldn't and if he did there is nothing we can do about it anyway).

shamardy replied:
We don't depend on the order publisher / maker as our source of truth ref. https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605#discussion_r2323208178

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2739,6 +2879,8 @@ impl Orderbook {
         let pair_root = order_pair_root_mut(&mut pubkey_state.trie_roots, &alb_ordered);
         let prev_root = *pair_root;
 
+        pubkey_state.pair_last_seen_local.insert(alb_ordered.clone(), now_sec());
```
shouldn't we similarly have such an update in `remove_order_trie_update`?

shamardy replied:
I don't want to refresh the maker liveness on cancelled orders, but I found something I missed and fixed it here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2605/commits/575d1d38da082529bf883dab2ee3b1f8aa1a33f0 where you will find a comment that describes this in a better way.

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2917,25 +3059,87 @@ impl Orderbook {
+                    log::debug!(
+                    "Ignoring PubkeyKeepAlive from {} due to stale pair {}: message.timestamp={} <= last_pair_timestamp={}",
+                    from_pubkey, alb_pair, message.timestamp, last_pair_timestamp
                 );
```
the closing parentheses should be indented

shamardy replied:
Yes, it's strange that fmt didn't catch this.

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2917,25 +3059,87 @@ impl Orderbook {
+                }
+
+                let last_pair_timestamp = {
+                    let pubkey_state = pubkey_state_mut(&mut self.pubkeys_state, from_pubkey);
```
we can put this at the top of the block before the for-loop.

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2917,25 +3059,87 @@ impl Orderbook {
+
+                let last_pair_timestamp = {
+                    let pubkey_state = pubkey_state_mut(&mut self.pubkeys_state, from_pubkey);
+                    *pubkey_state.latest_root_timestamp_by_pair.get(alb_pair).unwrap_or(&0)
```
`latest_root_timestamp_by_pair` is only read here to perform the check against `message.timestamp`.
this doesn't seem to be per-pair thing and we can have this per pubkey instead.

shamardy replied:
With a per pubkey timestamp, out‑of‑order keep‑alives or two pairs emitted with the same timestamp (what makers do currently in the code) will cause valid keepalives for some pairs to be rejected as stale. I actually hit this while working on this PR as I used a per pubkey timestamp at the start and a test failed iirc.

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -2917,25 +3059,87 @@ impl Orderbook {
+
+                if message.timestamp <= last_pair_timestamp {
+                    log::debug!(
+                    "Ignoring PubkeyKeepAlive from {} due to stale pair {}: message.timestamp={} <= last_pair_timestamp={}",
```
the error message should say `stale keepalive` rather than `stale pair`, i think.

shamardy replied:
Fixed

mariocynicys reviewed (commented):
another round.

shamardy committed:
review fixes: move `SyncFailure` comment and use `filter_map` instead of for loop in `build_pubkey_state_sync_request`

shamardy committed:
review fixes: clarify rate‑limiter comment

shamardy committed:
review fix: remove double iteration while avoiding overlapping borrows and extra allocations

shamardy committed:
review fixes: prefer `values().max().copied()`, use ASCII apostrophe

shamardy committed:
ordermatch: on cancel, advance per‑pair timestamp floor; drop liveness if a pair is emptied
- also remove an old commented out code for orderbook validation

shamardy committed:
review fixes: last few ones

shamardy commented:
@mariocynicys @dimxy this is ready for another review iteration

mariocynicys commented on mm2src/mm2_main/tests/mm2_tests/mm2_tests_inner.rs:
```diff
@@ -1870,7 +1869,7 @@ fn test_order_should_not_be_displayed_when_node_is_down() {
 
     block_on(mm_bob.stop()).unwrap();
-    thread::sleep(Duration::from_secs(6));
+    thread::sleep(Duration::from_secs(16));
```
this wait changed to 16 ...

mariocynicys commented on mm2src/mm2_main/tests/mm2_tests/mm2_tests_inner.rs:
```diff
@@ -1910,7 +1909,6 @@ fn test_own_orders_should_not_be_removed_from_orderbook() {
             "coins": coins,
             "i_am_seed": true,
             "rpc_password": "pass",
-            "maker_order_timeout": 5,
             "is_bootstrap_node": true
         }),
```
while the wait related to this test (line 1940) is still 6. is this the intended test?

shamardy replied:
right, this should be 16 to test that timing out doesn't affect own order

mariocynicys commented on mm2src/mm2_main/src/lp_ordermatch.rs:
```diff
@@ -310,50 +331,162 @@ fn process_trie_delta(
+        return Ok(());
+    }
+
+    // Query ONLY the origin peer.
```
let's replace `origin` with `propagator`?

mariocynicys reviewed (commented):
Thanks! Resolved the last comments from the previous PR (after re-reviewing knowing we deal with the propagating seednode and not the originator).

Couple of last nits/questions.

shamardy committed:
Merge remote-tracking branch 'origin/dev' into hotfix/sync-orderbook-fallback

shamardy committed:
review fixes: `test_own_orders_should_not_be_removed_from_orderbook` should wait 16 secs as this is the timeout for maker orders

mariocynicys reviewed (approved):
Thanks! LGTM!
